Tests/MongoSwiftTests/AuthTests.swift

```diff
@@ -0,0 +1,300 @@
+import Foundation
+import mongoc
+@testable import MongoSwift
+import Nimble
+import XCTest
+
+/// An extension adding accessors for a number of options that may be set on a `ConnectionString`.
+extension ConnectionString {
+    /// Returns the username if one was provided, otherwise nil.
+    private var username: String? {
+        guard let username = mongoc_uri_get_username(self._uri) else {
+            return nil
+        }
+        return String(cString: username)
+    }
+
+    /// Returns the password if one was provided, otherwise nil.
+    private var password: String? {
+        guard let pw = mongoc_uri_get_password(self._uri) else {
+            return nil
+        }
+        return String(cString: pw)
+    }
+
+    /// Returns the auth database if one was provided, otherwise nil.
+    private var authSource: String? {
+        guard let source = mongoc_uri_get_auth_source(self._uri) else {
+            return nil
+        }
+        return String(cString: source)
+    }
+
+    /// Returns the auth mechanism if one was provided, otherwise nil.
+    private var authMechanism: AuthMechanism? {
+        guard let mechanism = mongoc_uri_get_auth_mechanism(self._uri) else {
+            return nil
+        }
+        let str = String(cString: mechanism)
+        return AuthMechanism(rawValue: str)
+    }
+
+    /// Returns a document containing the auth mechanism properties if any were provided, otherwise nil.
+    private var authMechanismProperties: Document? {
+        var props = bson_t()
+        return withUnsafeMutablePointer(to: &props) { propsPtr in
+            let opaquePtr = OpaquePointer(propsPtr)
+            guard mongoc_uri_get_mechanism_properties(self._uri, opaquePtr) else {
+                return nil
+            }
+            /// This copy should not be returned directly as its only guaranteed valid for as long as the
+            /// `mongoc_uri_t`, as `props` was statically initialized from data stored in the URI and may contain
+            /// pointers that will be invalidated once the URI is.
+            let copy = Document(copying: opaquePtr)
+
+            return copy.mapValues { value in
+                // mongoc returns boolean options e.g. CANONICALIZE_HOSTNAME as strings, but they are booleans in the
+                // spec test file.
+                guard let str = value as? String else {
+                    return value
+                }
+                switch str {
+                case "true":
+                    return true
+                case "false":
+                    return false
+                default:
+                    return str
+                }
+            }
+        }
+    }
+
+    /// Returns the credential configured on this URI. Will be empty if no options are set.
+    fileprivate var credential: Credential {
```

Comment:
I imagine that eventually we would want to incorporate this into `ClientOptions`, as well as all other options supported by the URI. I filed SWIFT-640 for the specific auth options case, but perhaps we would want a "URI/ClientOptions Parity" epic for the entire project.

Perhaps add a TODO here or somewhere else in this file for updating the tests to go over the `ClientOptions` path too?

Reply:
I added a TODO by the prose tests, since those are what the spec says should be tested via both URI and options struct.

Reply:
also opened [SWIFT-641](https://jira.mongodb.org/browse/SWIFT-641) with your proposed epic